pallets/connectors-gateway/connectors-gateway-routers/src/axelar_evm.rs

```diff
@@ -0,0 +1,217 @@
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+use cfg_traits::connectors::Codec;
+use codec::{Decode, Encode, MaxEncodedLen};
+use ethabi::{Contract, Function, Param, ParamType, Token};
+use frame_support::dispatch::{DispatchError, DispatchResult, RawOrigin};
+use scale_info::{
+	prelude::string::{String, ToString},
+	TypeInfo,
+};
+use sp_core::{H160, U256};
+use sp_std::{collections::btree_map::BTreeMap, marker::PhantomData, vec, vec::Vec};
+
+use crate::{AccountIdOf, MessageOf};
+
+#[derive(Debug, Encode, Decode, Clone, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub struct AxelarEVMRouter<T>
+where
+	T: frame_system::Config + pallet_connectors_gateway::Config + pallet_evm::Config,
+{
+	pub domain: EVMDomain,
+	pub _phantom: PhantomData<T>,
+}
+
+#[derive(Debug, Encode, Decode, Clone, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub struct EVMDomain {
+	pub chain: EVMChain,
+	pub axelar_contract_address: H160,
+	pub connectors_contract_address: H160,
+	pub fee_values: FeeValues,
+}
+
+#[derive(Debug, Encode, Decode, Clone, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub struct FeeValues {
+	pub value: U256,
+	pub gas_limit: u64,
+	pub max_fee_per_gas: U256,
+	pub max_priority_fee_per_gas: Option<U256>,
+}
+
+#[derive(Debug, Encode, Decode, Clone, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub enum EVMChain {
+	Ethereum,
+}
+
+/// Required due to the naming convention defined by Axelar here:
+/// https://docs.axelar.dev/dev/reference/mainnet-chain-names
+impl ToString for EVMChain {
+	fn to_string(&self) -> String {
+		match self {
+			EVMChain::Ethereum => "Ethereum".to_string(),
+		}
+	}
+}
+
+impl<T> AxelarEVMRouter<T>
+where
+	T: frame_system::Config + pallet_connectors_gateway::Config + pallet_evm::Config,
+	T::AccountId: AsRef<[u8; 32]>,
+{
+	pub fn do_send(&self, sender: AccountIdOf<T>, msg: MessageOf<T>) -> DispatchResult {
+		let eth_msg = self.get_eth_msg(msg).map_err(|e| DispatchError::Other(e))?;
+
+		// Use the same conversion as the one used in `EnsureAddressTruncated`.
+		let sender_evm_address = H160::from_slice(&sender.as_ref()[0..20]);
+
+		// TODO(cdamian): Do we need extra account related logic, prior to calling this
+		// router, given the following?
+		//
+		// `pallet_evm::runner::stack::Runner::validate` will convert
+		// `sender_evm_address` into a `T::AccountID` and then attempt to return the
+		// nonce and balance for this account, the account mutation flow being:
+		//
+		// AxelarEVMRouter:
+		//
+		// Sender T::AccountId -> Sender H160
+		//
+		// pallet_evm::Pallet::account_basic:
+		//
+		// Sender H160 -> Unknown T::AccountId
+		//
+		// This will most likely mean that the information for
+		// the unknown account will be empty i.e. 0 nonce and 0 balance.
+		pallet_evm::Pallet::<T>::call(
```

Comment:
tl;dr - we probably need to have a broader sync about requirements for calling EVM from substrate code and how we want to deal with gas, weights, and fees - there are lots of complications here.

Normally, the EVM wants to handle charging transaction fees for us. This is fine, until it's not. It's kinda not fine here.

* As noted, the AccountID doesn't round-trip cleanly.
* We need to account for the weight of the EVM work. Normally this is done by a `PostDispatchInfo` with `pays_fee: false`, since the EVM has already charged the fees. But in this case we have additional work in the top-level extrinsic that needs to be paid for.

We have a few ways of dealing with this:
* We can continue to use `pays_fee: false`, by letting the EVM do its thing. We'd also need to manually handle charging the fee for the non-EVM parts of any extrinsic which reaches this code path. This will require making accounts round-trip so that the EVM can charge the correct top-level AccountID for this call.
* We can pay fees normally with a standard `PostDispatchInfo`. This means making the EVM `OnChargeTransaction` handle the case where it should lock fees but not finalize them, letting the total weight be handled at the very end.
  * I think for this to work correctly we would need to adjust fee calculations so that EVM gas prices are tied to substrate weights by using `pallet_transaction_payment` instead of a separate EVM price computation. Otherwise the EVM could lock too few tokens for payment if gas prices are cheaper than substrate weight prices.
* We could reimplement the parts of the `Runner` that are needed to nicely call from within an extrinsic, bypassing all EVM fee logic entirely and just getting back a weight that we can add to the rest of our tranasaction.

All of this gets more complicated if we expect the call to come from an extrinsic which might have been triggered by an Ethereum call in the first place, because now we're dealing with gas->weight->gas conversions and possibly fees being charged in multiple places. I haven't dived into those codepaths yet at all.

Reply:
As for the best way to actually _invoke_ the call: I belive it'll be `pallet_ethereum::Pallet<T>::execute(origin, transaction)`

Reply:
Thank you for the input, I think it makes sense to refrain from making any changes until we figure out how to deal with the fee issue that you described. Initially I wanted to change to `pallet_ethereum::Pallet<T>::execute` but since that might also not suffice I'll hold it until we clarify that and focus on the moonbeam router tests for the time being.

cc @mustermeiszer Please check this and let us know if we should sync more on this topic.

Reply:
Another complication: since this is for axelar, we probably need a transaction + events to appear on the Ethereum block. So maybe this needs to look more like how Moonbeam does `ethereum_xcm`?

Regardless, I still think it'd be good for us to fully understand the requirements here so we can sort out the right solution

Reply:
@branan I think we can get away with charging the fees twice. Once for the actual calley, the user who wants to use the bridge, and then secondly on the EVM side charging the pallet-account that we accept on the ethereum side. 

All we need to do is (re-)fund the pallet-account form the treasury. Either automatically or via democracy.

Reply:
@cdamian I would for now move forward with this! We should keep in mind that we need to look into fee managmanet later on. WDYT>

Reply:
@mustermeiszer I agree.

Reply:
Resolving this for now, let's start another discussion on this topic if needed.